libV2/tasks/IngestShard.js

```diff
@@ -0,0 +1,54 @@
+const assert = require('assert');
+const BaseTask = require('./BaseTask');
+const { UtapiMetric } = require('../models');
+const config = require('../config');
+const { LoggerContext, shardFromTimestamp } = require('..//utils');
+const { shardIngestLagSecs } = require('../constants');
+
+const logger = new LoggerContext({
+    module: 'IngestShard',
+});
+
+class IngestShardTask extends BaseTask {
+    constructor(...options) {
+        super(...options);
+        this._defaultSchedule = config.ingestionSchedule;
+    }
+
+    async _execute(timestamp) {
+        const endShard = shardFromTimestamp(timestamp - (shardIngestLagSecs * 1000000));
+        logger.debug('ingesting shards', { endShard });
+
+        const available = await this._cache.getShards();
+        const toIngest = available
+            .map(key => key.split(':')[2])
+            .filter(key => parseInt(key, 10) <= endShard);
+
+        if (!toIngest.length) {
+            logger.debug('no shard available to ingest');
+            return;
+        }
+
+        await Promise.all(toIngest.map(
+            async shard => {
+                if (await this._cache.shardExists(shard)) {
+                    const metrics = await this._cache.getMetricsForShard(shard);
+                    if (metrics.length) {
+                        logger.info(`Ingesting ${metrics.length} events from shard`, { shard });
+                        const records = metrics
+                            .map(m => new UtapiMetric(JSON.parse(m)));
+                        const status = await this._warp10.ingest('utapi.event', records);
+                        assert.strictEqual(status, records.length);
```

Comment:
I'm personally ok with using assertions in the code (some people may disagree) but make sure it's really an assertion and not an error case to handle, so by design `warp10.ingest` must return as many statuses as records, whatever happens. Did not check this but wanted to make the point anyway.

Reply:
I was using `assert` as a more compact way to throw an `Error` here, and as plus you get a nicely formatted error message

Reply:
Ah ok, why not, even though we tend to use assert e.g. for config parsing to fail early, I think it better to explicitly throw errors and reserve asserts for programming logic errors, but up to you I'm good with that too.